cpp/src/io/csv/datetime_parser.cuh

```diff
@@ -306,58 +330,88 @@ bool extractTime(char *data, int sIdx, int eIdx, int *hour, int *minute, int *se
 	return true;
 }
 
+// User-defined literals to clarify numbers and units for time calculation
+__host__ __device__
+constexpr uint32_t operator "" _days(unsigned long long int days) {
+	return days;
+}
+__host__ __device__
+constexpr uint32_t operator "" _erasInDays(unsigned long long int eras) {
+	// multiply by number of days within an era (400 year span)
+	return eras * 146097_days;
+}
+__host__ __device__
+constexpr uint32_t operator "" _years(unsigned long long int years) {
+	return years;
+}
+__host__ __device__
+constexpr uint32_t operator "" _erasInYears(unsigned long long int eras) {
+	return (eras * 1_erasInDays) / 365_days;
+}
 
 /**
- * @brief compute number of days since epoch
+ * @brief Compute number of days since "March 1, 0000", given a date
  *
- * This function takes year, month, and day and return the
- * number of days since epcoh as a int32
+ * This function takes year, month, and day and returns the number of days
+ * since the baseline which is taken as 0000-03-01. This value is chosen as the
+ * origin for ease of calculation (now February becomes the last month).
  *
  * @param[in] year
  * @param[in] month
  * @param[in] day
  *
- * @return days since epoch
- *
+ * @return days since March 1, 0000
  */
 __host__ __device__
-gdf_date32 daysSinceEpoch(int year, int month, int day)  {
-
-	static unsigned short days[12] = {0,  31,  59,  90, 120, 151, 181, 212, 243, 273, 304, 334};
-
-	// years since epoch
-	int ye = year - 1970;
-
-	// 1972 was a leap year, so how many have occurred between date and 1972? Do not include this year in count
-	int lpy = ((year - 1972 - 1) / 4);
-
-	// compute days since epoch
-	gdf_date32 days_e = (ye * 365) + lpy;
-
-	// is this a leap year?
-	if ( year % 4 == 0 && month > 2)
-		days_e++;
+constexpr int32_t daysSinceBaseline(int year, int month, int day) {
+
+    // More details of this formula are located in cuDF datetime_ops
+    // In brief, the calculation is split over several components:
+    //     era: a 400 year range, where the date cycle repeats exactly
+    //     yoe: year within the 400 range of an era
+    //     doy: day within the 364 range of a year
+    //     doe: exact day within the whole era
+    // The months are shifted so that March is the starting month and February
+    // (possible leap day in it) is the last month for the linear calculation
+    year -= (month <= 2) ? 1 : 0;
+
+    const int32_t era = (year >= 0 ? year : year - 399_years) / 1_erasInYears;
```

Comment:
Before I merge, I want to make sure you are definitely intending integer division throughout this computation?

Reply:
Yes, the original whitepaper intends it to be integer division. We could simplify it if we didn't want to handle negative years but I wanted to keep the algorithm mostly intact.

Reply:
OK!